plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+
+  -- 2nd-pass functions
+  ---------------------
+
+  -- | visit each term to apply the float transformation
+  -- OPTIMIZE: stop descending when state==mempty?
+  goTerm :: Int -- ^ current depth
+         -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+         -> Term tyname name uni a
+         -> State [S.Set PLC.Unique] (Term tyname name uni a)
+  goTerm curDepth searchTable = \case
+      -- we are only interested in lambdas/Lambdas
+      LamAbs a n ty tBody -> LamAbs a n ty <$> goAbs curDepth searchTable n tBody
+      TyAbs a n k tBody  -> TyAbs a n k <$> goAbs curDepth searchTable n tBody
+      -- descend otherwise to apply the transformations to subterms
+      t -> termSubterms (goTerm curDepth searchTable) t
+
+  -- | If a lambda/Lambda is found, the current location is updated (depth+lamUnique) and try to float its body
+  goAbs :: PLC.HasUnique b b'
+        => Int -- ^ current depth
+        -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+        -> b                  -- ^ lambda/Lambda's unique
+        -> Term tyname name uni a -- ^ lambda/Lambda's body
+        -> State [S.Set PLC.Unique] (Term tyname name uni a)
+  goAbs oldDepth floatData n tBody =
+    let newLam = LamDep (oldDepth+1) (n^.PLC.theUnique)
+    in goFloat newLam tBody floatData
+
+  -- | We are currently INSIDE (exactly under) a lambda/Lambda body/Top (Top if we are right at the start of the algorithm)
+  -- We try to see if we have some lets to float here based on our 1st-pass-table data (searchTable).
+  goFloat :: Rank -- ^ the rank/location of the lambda above that has this body
+         -> Term tyname name uni a                           -- ^ the body term
+         -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+         -> State [S.Set PLC.Unique] (Term tyname name uni a)                           -- ^ the transformed body term
+  goFloat aboveLamLoc tBody floatData  =
+   -- look for the next smallest depth remaining to place
+   case IM.minViewWithKey floatData of
+     Nothing -> pure tBody -- nothing left to float
+     Just ((searchingForDepth, searchingForLams_Lets), restFloatData) ->
+      let curDepth = aboveLamLoc^.depth
+      in case curDepth `compare` searchingForDepth of
+      -- the minimum next depth we are looking for, is not this one, so just descend with the whole floatdata
+      LT -> goTerm curDepth floatData tBody
+      -- found depth, see if our lambda above is a lambda we are interested in (to float some lets)
+      EQ ->
+        -- transform the lam-body, passing the rest floatdata
+        let tBody' = goTerm curDepth restFloatData tBody
+        in case M.lookup aboveLamLoc searchingForLams_Lets of
+             -- the lambda above-this-body has some let-rhses to insert (float) here.
+             Just lets -> genLets lets restFloatData =<< tBody'
```

Comment:
Hm... So we look a single `Rank` up in a map and then throw away the entire map? That looks strange.

Reply:
Ah, got it: we throw away the map here, but we'll do a similar look-up in another lambda at the same depth.